crates/oxc_linter/src/rules/promise/no_return_wrap.rs

```diff
@@ -0,0 +1,299 @@
+    OxcDiagnostic::warn("Expected throw instead of Promise.reject").with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoReturnWrap(Box<NoReturnWrapOptions>);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoReturnWrapOptions {
+    pub allow_reject: bool,
+}
+
+// This documentation and the following test cases are based on the original
+// implementation from eslint-plugin-promise and are licensed under the ISC License.
+//
+// Copyright (c) 2020, Jamund Ferguson
+//
+// See the LICENSE file in the root of this repository for more information.
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Ensure that inside a then() or a catch() we always return or throw a raw value instead of
+    /// wrapping in Promise.resolve or Promise.reject
+    ///
+    /// ### Why is this bad?
+    ///
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// myPromise.then(function (val) {
+    ///     return Promise.resolve(val * 2)
+    /// })
+    /// myPromise.then(function (val) {
+    ///     return Promise.reject('bad thing')
+    /// })
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// myPromise.then(function (val) {
+    ///   return val * 2
+    ///   })
+    /// myPromise.then(function (val) {
+    ///     throw 'bad thing'
+    /// })
+    /// ```
+    NoReturnWrap,
+    style,
+);
+
+impl Rule for NoReturnWrap {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config = value.get(0);
+
+        let allow_reject = config
+            .and_then(|c| c.get("allowReject"))
+            .and_then(serde_json::Value::as_bool)
+            .map_or(false, |v| v);
+
+        Self(Box::new(NoReturnWrapOptions { allow_reject }))
+    }
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+        if call_expr.callee.as_member_expression().is_some()
+            && is_promise(call_expr).is_some()
+            && is_in_promise(node.id(), ctx)
+        {
+            if matches!(call_expr.callee_name(), Some("resolve")) {
+                ctx.diagnostic(no_return_wrap_diagnostic_resolve(call_expr.span));
+            }
+            if !self.0.allow_reject && matches!(call_expr.callee_name(), Some("reject")) {
+                ctx.diagnostic(no_return_wrap_diagnostic_reject(call_expr.span));
+            }
+        }
+    }
+}
+
+fn is_in_promise(id: NodeId, ctx: &LintContext) -> bool {
+    // find the innermost function node
+    let mut function_node = ctx.nodes().ancestors(id).find(|node| {
+        matches!(node.kind(), AstKind::ArrowFunctionExpression(_) | oxc_ast::AstKind::Function(_))
+    });
+
+    // move up to the parent node while the function is re-bound with "bind"
+    while let Some(func_node) = function_node {
```

Comment:
The function has multiple responsibilities, such as searching for `function` nodes, processing `bind` nodes, and determining if a node is a `Promise`, so I attempted to refactor it.
However, due to lifetime issues, I decided not to proceed with the split. Ultimately, changing the run function’s signature as shown below would have been necessary:
```diff
- run<'a>(..., ctx: & LintContext<'a>)
+ run<'a>(..., ctx: &'a LintContext<'a>)

```